evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -8,28 +14,146 @@
         self.default_sender_address = default_privkey.public_key.to_canonical_address()
         super().__init__(*args, **kwargs)
 
+    def setup_collation_added_filter(self, shard_id):
+        shard_id_topic = "0x" + shard_id.to_bytes(32, byteorder='big').hex()
+        self.collation_added_filter[shard_id] = self.web3.eth.filter({
+            'address': self.address,
+            'topics': [
+                self.collation_added_topic,
+                shard_id_topic,
+            ],
+        })
+        self.new_collation_added_logs[shard_id] = []
+        self.unchecked_collation_added_logs[shard_id] = []
+        self.current_checking_score[shard_id] = None
+
+    @to_dict
+    def parse_collation_added_data(self, data):
+        score = big_endian_to_int(data[-32:])
+        is_new_head = bool(big_endian_to_int(data[-64:-32]))
+        header_bytes = data[:-64]
+        # [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, num, bytes]
+        sedes = rlp.sedes.List([
+            rlp.sedes.big_endian_int,
+            rlp.sedes.big_endian_int,
+            hash32,
+            hash32,
+            hash32,
+            address,
+            hash32,
+            hash32,
+            rlp.sedes.big_endian_int,
+            rlp.sedes.binary,
+        ])
+        header_values = rlp.decode(header_bytes, sedes=sedes)
+        yield 'header', header_values
+        yield 'is_new_head', is_new_head
+        yield 'score', score
+
+    @to_tuple
+    def _get_new_logs(self, shard_id):
+        # use `get_new_entries` over `get_all_entries`
+        #   1. Prevent from the increasing size of logs
+        #   2. Leave the efforts maintaining `new_logs` in RPC servers
+        if shard_id not in self.collation_added_filter:
+            raise FilterNotFound(
+                "CollationAdded filter haven't been set up in shard {}".format(shard_id)
+            )
+        new_logs = self.collation_added_filter[shard_id].get_new_entries()
+        for log in new_logs:
+            yield pipe(
+                log['data'],
+                decode_hex,
+                self.parse_collation_added_data,
+            )
+
+    def get_next_log(self, shard_id):
+        new_logs = self._get_new_logs(shard_id)
+        self.new_collation_added_logs[shard_id] += new_logs
+        if self.new_collation_added_logs[shard_id] == []:
+            raise NextLogUnavailable("No more next logs")
+        return self.new_collation_added_logs[shard_id].pop()
+
+    def fetch_candidate_head(self, shard_id):
+        # Try to return a log that has the score that we are checking for,
+        # checking in order of oldest to most recent.
+        if shard_id not in self.collation_added_filter:
+            raise FilterNotFound(
+                "CollationAdded filter haven't been set up in shard {}".format(shard_id)
+            )
+        for i in range(len(self.unchecked_collation_added_logs[shard_id]) - 1, -1, -1):
```

Comment:
I *think* this can be simplified with:

```
for log_entry in reversed(self.unchecked_collation_added_logs[shard_id]):
    if log_entry == ...:
        ...
```

Reply:
Thank you for pointing it out. Have changed it.